projects/angular/src/lib/maskito.directive.ts

```diff
@@ -30,16 +30,16 @@ export class MaskitoDirective implements OnDestroy, OnChanges {
 
     @Input()
     maskitoElement: MaskitoElementPredicate = e =>
-        e.querySelector('input,textarea') ||
+        e.querySelector<HTMLInputElement | HTMLTextAreaElement>('input,textarea') ||
         (e as HTMLInputElement | HTMLTextAreaElement);
 
-    ngOnChanges(): void {
+    async ngOnChanges(): Promise<void> {
         this.maskedElement?.destroy();
-        this.ngZone.runOutsideAngular(() => {
-            this.maskedElement = new Maskito(
-                this.maskitoElement(this.elementRef.nativeElement),
-                this.maskito,
-            );
+
+        await this.ngZone.runOutsideAngular(async () => {
+            const predicate = await this.maskitoElement(this.elementRef.nativeElement);
+
+            this.maskedElement = new Maskito(predicate, this.maskito);
```

Comment:
We gotta be wary of race conditions.

What happens if async predicate function will return value too long?

For example:



1. Developer pass async `predicate#1`
2. ngOnChanges fires and execute `predicate#1`
3. `predicate#1` will return value only in 5 seconds
4. Developer pass new async `predicate#2` (before `predicate#1` is complete)
5. ngOnChanges fires and execute `predicate#2`
6. `predicate#2` instantly completes and create `Mask#2`
7. `predicate#1` completes and create `Mask#1`

Reply:
For `@maskito/angular` package we could wrap it with `Observable` + use `switchMap`.

Reply:
I don't know what we should do for `@maskito/react`.

@nextZed Could you suggest solution to process this case in **React** way ?

Reply:
I have an idea, what if we save predicate to a const before resolving it and then check if it remained the same? Pseudo code:
```ts
const predicate = this.predicate
const result = async predicate(element);

if (this.predicate !== predicate) {
  return;
}

this.maskedElement = new Maskito(result, options);
```

Reply:
Addressed here: https://github.com/Tinkoff/maskito/pull/316/commits/4dea275d78d7d2d86c026056df8c09a6096aaa84 (removed the VSCode settings file in a later commit). 

Based on the feedback here: https://github.com/Tinkoff/maskito/pull/316#discussion_r1203629086 I implemented this in the Angular implementation only and reverted the react changes to remain synchronous. I'll do more discovery into the usage impact with my use case of Ionic React (web components) and the Maskito React hook.